Terminal.Gui/Application.cs

```diff
@@ -658,6 +647,33 @@ public static void Run (Toplevel view, Func<Exception, bool> errorHandler = null
         }
     }
 
+    private static void EnsureValidInitialization (Toplevel top, ConsoleDriver driver)
```

Comment:
It's called by the `Run<T>` and the `Run(v)`. See it again.

Reply:
But there is also many calls to `Run(v)` without using the `Run<T>()`, where we need to ensure the driver was initialized.

Reply:
Suggestion:

- Change `Run(View)` and `Run<T>()` to return `Toplefvel`
- Change the API docs of `Run<T>()` to say "The caller is responsible for disposing the Toplevel created by this method."
- Change the API docs of `Run (View)` to say "The returned Toplevel is the same as the one passed in as the toplevel parameter. The caller is responsible for disposing this object."
- In both cases the API docs should say "This method (via Begin) sets Application.Top to the toplevel passed to this method (or created by it). Note that nested calls to Application.Run may change Application.Top, thus callers should hold a different reference for disposal."

Consider deleting ` public static void Run (Func<Exception, bool> errorHandler = null, ConsoleDriver driver = null)`. If you don't delete it, at least make it return the `Toplevel` that gets created and ensure its API docs cover the points above.

Reply:
Unless you want to make the Run(v) private which I don't advice.

Reply:
I think the most common use case, that most of our sample code/Scenarios should illustrate is:

```cs
Application.Init ();
var app = Application.Run<MyAppView> ();
app.Dispose ();
Application.Shutdown ();
```

I used "app" here because when we make it so any View can be `Run`, the overrall concept of the "top level view" sorta goes away. 
Note

Reply:
> Unless you want to make the Run(v) private which I don't advice.

Ahhh, you're worried about the fact that `Run<T>` calls `Init` if it was not already called.

Why can't `Run(v)` do the same thing?

Reply:
> Change Run(View) and Run<T>() to return Toplefvel

I really don't see the advantage to Run return a Toplevel. That way the app will be running before you add views. That is a waste of resources.

> Why can't Run(v) do the same thing?

It's already doing the same.

Reply:
> > Change Run(View) and Run() to return Toplefvel
> 
> I really don't see the advantage to Run return a Toplevel. That way the app will be running before you add views. That is a waste of resources.

If you mean " I really don't see the advantage to `Run(r)`... but I *do* see why it makes sense for `Run()` and `Run<T>()`..." then cool! 

I suggested  making all 3 of them return `Toplevel` for consistency, but I think it's fine without it too.

Reply:
> I suggested making all 3 of them return `Toplevel` for consistency, but I think it's fine without it too.

Normally WPF, Windows Forms only use void. Imagine you want to big toplevel with many views and subviews. Only after you call Run() or Run<T>() you will add views with the IsInitialized as true, where we'll have a bunch of issues because the view is already initialized, that we'll don't have when we create them before run. I hope you understand this.